WIP: Generated using the latest features from MoveIt

tylerjw opened the pull request:
# Description

This was created by generating using the master branch of MoveIt.  There are a handful of files that were not generated (i.e. `config/panda_arm_hand.srdf.xacro`), I'd like to document how they are created and what their use is in the README.  

Lastly, I chose two new poses that are as close to the original ones as I could make and stay out of collision.  I didn't include the `transport` pose because I was not able to make that pose and stay out of collision.  I've included images of each pose here:
![panda_extended](https://user-images.githubusercontent.com/2490389/94869641-f910b700-0402-11eb-9a38-fce9a5d2a0ce.png)
![panda_ready](https://user-images.githubusercontent.com/2490389/94869643-f9a94d80-0402-11eb-9bc4-9b9e902cc56b.png)
![panda_transport](https://user-images.githubusercontent.com/2490389/94869645-f9a94d80-0402-11eb-81fd-630d365a582b.png)

Lastly, this should be tested with moveit_tutorials to make sure it doesn't break any of those.

## Notes to reviewers

Here are the things that need to be done before this is ready:
* [ ] Identify and document changes made after generating from MSA
* [ ] Discuss new poses
* [ ] Test with moveit_tutorials

tylerjw committed:
Generated using the latest features from MoveIt

v4hn commented:
The `transport` pose is the configuration used to ship the robot in its original box.
That's the main reason why it was part of the configuration.

Yes, it's in collision with the current model and this was discussed somewhere before. The MoveIt-compatible trajectory controller can't move the robot to this pose because the internal self-collision check will fail as well. Apparently they have a separate internal routine to move to this pose.

tylerjw commented:
I checked through the commits since the last release and this only backs out one change (there is a PR to add the transport pose to one of the files).

This brings us to a point of discussion.  There are two sorts of files in this repo that are not generated by moveit_setup_assistant.  The first is xacro files that were copied versions of the old robot description.  I believe these can be safely removed.  

The next are files that support specific tutorials and were manually added here.  An example of that is the `lerp_planning_pipeline.launch.xml` and related changes.  I'm not sure what the right approach is for these sorts of files.  I think it makes sense to document what these files are so that in the future if someone needs to generate this package to update it they know what to do about those files.

That or we should move these launch and config files out of this repo and into the tutorials repo to go along side their respective tutorials.  A third option would be to extend the setup assistant to generate these files if they are actually general-purpose launch files that would be useful for other robot setups.

@rhaschke @v4hn do you have thoughts on how we should best maintain these manually added files?  I am working through identifying them and trying to validate the functionality of this PR with tutorials.

tylerjw commented:
Here is a list of the tutorials on master of moveit_tutorials with my notes.  I'll add notes as I work through each one comparing it to the current version of panda_moveit_config.

* [x] Getting Started
* [x] MoveIt Quickstart in RViz
* [ ] Move Group C++ Interface
* [ ] Move Group Python Interface
* [ ] MoveIt Commander Scripting
* [ ] Robot Model and Robot State
* [ ] Planning Scene
* [ ] Planning Scene Monitor
* [ ] Planning Scene ROS API
* [ ] Motion Planning API
* [ ] Motion Planning Pipeline
* [ ] Creating MoveIt Plugins
* [ ] Visualizing Collisions
* [ ] Time Parameterization
* [ ] Planning with approximated Constraint Manifolds
* [ ] Pick and Place
* [ ] MoveIt Grasps
* [ ] MoveIt Task Constructor
* [ ] MoveIt Deep Grasps
* [ ] Subframes
* [ ] MoveItCpp Tutorial
* [ ] MoveIt Setup Assistant
* [ ] URDF and SRDF
* [ ] Low Level Controllers
* [ ] Preception Pipeline Tutorial
* [ ] IKFast Kinematics Solver
* [ ] TRACK-IK Kinematics Solver
* [ ] Kinematics Configuration
* [ ] Custom Constraint Samplers
* [ ] OMPL Planner
* [ ] CHOMP Planner
* [ ] STOMP Planner
* [ ] TrajOpt Planner
* [ ] Planning Adapter Tutorials
* [ ] Joystick Control Teleoperation
* [ ] Realtime Arm Servoing
* [ ] Benchmarking
* [ ] Integration/Unit Tests

tylerjw committed:
restore rviz file

v4hn commented:
> The first is xacro files that were copied versions of the old robot description.
> I believe these can be safely removed.

This is not the "old robot description", but the srdf file that is split in multiple sub-files. If you drop the ability to load the arm without gripper from the updated package, they should be removed. Some users might be interested in using the arm with other endeffectors, but they should usually generate their own moveit config anyway. So unless the distinction is required for some tutorial, I assent to revert to a single srdf file.

> A third option would be to extend the setup assistant to generate these files if they are actually general-purpose launch files that would be useful for other robot setups.

If some files are general-purpose, they should be added to the templates.
The LERP planner is created in the tutorial though, so I would just keep it in the panda config with an appropriate comment at the top of the file.

> That or we should move these launch and config files out of this repo and into the tutorials repo to go alongside their respective tutorials. 

The `planning_pipeline` "plugin-like" structure does not allow for this and we discussed some time ago in some issue whether we want to extend it to do so. At least my personal opinion was that it does not make sense to add plugin mechanisms to launch files and people can directly include them in their own launch files.
However, I'm not sure how much overhead that would create for the tutorial's launch file, which should definitely be as simple as possible to support new users.

tylerjw commented:
> This is not the "old robot description", but the srdf file that is split in multiple sub-files. If you drop the ability to load the arm without gripper from the updated package, they should be removed. Some users might be interested in using the arm with other endeffectors, but they should usually generate their own moveit config anyway. So unless the distinction is required for some tutorial, I assent to revert to a single srdf file.

I believe this functionality is now provided by the description files in the upstream franka package.  From what I could tell our versions of these do not differ in functionality from the official ones now.  I have yet to find a tutorial that used them, or tested loading without the gripper using the upstream description files.

tylerjw commented:
Today I went through and investigates some of the easier to test tutorials.  I'll need to work through the rest of them but unfortunatly I found some that don't work with the old config and the master branch of MoveIt on Noetic.  Some of these might be easy fixes but I'd don't believe they should block this change and should hopefully be unaffected by this change.

v4hn commented:
> > This is not the "old robot description", but the srdf file that is split in multiple sub-files.
> 
> I believe this functionality is now provided by the description files in the upstream franka package. From what I could tell our versions of these do not differ in functionality from the official ones now. I have yet to find a tutorial that used them, or tested loading without the gripper using the upstream description files.

Are we talking about the same files? [hand.xacro in this repository](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/hand.xacro#L9) provides **SRDF** specifications, [hand.xacro in franka_ros](https://github.com/frankaemika/franka_ros/blob/kinetic-devel/franka_description/robots/hand.xacro#L12) provides **URDF** specs. These files do not have anything in common aside from their name.

tylerjw commented:
> Are we talking about the same files? [hand.xacro in this repository](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/hand.xacro#L9) provides **SRDF** specifications, [hand.xacro in franka_ros](https://github.com/frankaemika/franka_ros/blob/kinetic-devel/franka_description/robots/hand.xacro#L12) provides **URDF** specs. These files do not have anything in common aside from their name.

I didn't realize this.  I figured they provided the same utility.  My reasoning was based on searching for files that used these various files in the panda_moveit_config repo and in all cases it seemed like they were referring to the one in franka_description.  I'll put those back.

tylerjw committed:
Sync allowed collisions in panda_arm.xacro with panda.srdf

henningkayser commented on config/panda_arm.xacro:
```diff
@@ -11,31 +11,22 @@
+        <joint name="panda_joint4" value="-2.356"/>
+        <joint name="panda_joint5" value="0"/>
+        <joint name="panda_joint6" value="1.571"/>
+        <joint name="panda_joint7" value="-0.7855"/>
```
Why are the default poses changing? They shouldn't be affected by running MSA.

tylerjw replied:
for the upstream collision scene

henningkayser commented on launch/demo.launch:
```diff
@@ -1,39 +1,40 @@
+  <node name="joint_state_publisher" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" if="$(arg use_gui)">
+    <rosparam param="source_list">[move_group/fake_controller_joint_states]</rosparam>
   </node>
-  <node name="joint_state_desired_publisher" pkg="topic_tools" type="relay" args="joint_states joint_states_desired" />
```
Not sure if we should really remove this

henningkayser reviewed (commented):
Overall, I really think it makes a lot of sense to keep configs up-to-date with the MSA, imo it's just strainge if this is not possible. I'm just not sure what setups we would break by doing this. Of course we should keep modifications like with the srdf files and default poses consistent.

JafarAbdi committed:
Sync with MoveIt master MSA

JafarAbdi committed:
Disable collision between panda_link8 & panda_link6 + panda_link7

JafarAbdi commented:
Note that I disable the collision between panda_link8 & panda_link6 + panda_link7 otherwise hand movement is very restricted causing some of the tutorials to not work

davetcoleman commented:
This seems overdue for a merge, @tylerjw can we push this in?

tylerjw commented:
@davetcoleman I abandoned this because I didn't have the time to finish it.  There are a handful of tutorials it needs to be tested with and either this or the tutorials need to be fixed.  I've left documentation of all of that above.

davetcoleman commented:
@tylerjw by "abandoned" you mean set it aside, correct? You would support someone else taking this on, as its still important right?

tylerjw commented:
I mean that I stopped working on it.  I have no issue with someone else taking it up.  I think that person will need to have a good understanding of the uses of this package (probably more than me) and all the tutorials or they risk getting stuck just as I did.

rickstaa commented:
@tylerjw I made some changes that make sure that the `panda_control_moveit_rviz.launch` file can be used again to control the real robot using moveit. I created a pull request on your branch https://github.com/tylerjw/panda_moveit_config/pull/1.

rickstaa commented:
@tylerjw First of all thanks a lot for creating this pull request. @davetcoleman I'm happy to help. I made some changes to ensure that the `panda_control_moveit_rviz.launch` file can be used again to control the real robot using move it. In addition, I created a pull request on your branch https://github.com/tylerjw/panda_moveit_config/pull/1. I will create an additional pull requests, if I run into more errors while using your branch.

rickstaa committed:
Adds panda_control_moveit_rviz.launch fixes

rickstaa commented:
@tylerjw I updated the robot_description xacro url since the upstream was changed (see https://github.com/tylerjw/panda_moveit_config/pull/3).

rickstaa committed:
Updates robot_description xacro url because of upstream changes (#3)

This commit makes sure the robot_description is loaded from the right
url. This was done since
https://github.com/frankaemika/franka_ros/commit/3438d64f3499a22d297465f267854d9037e88996
in the franka_ros package combines the hand and arm xacro files.

rickstaa commented:
@tylerjw Thanks for merging the last pull request I created a new pull request with two fixes that are necessary (see https://github.com/tylerjw/panda_moveit_config/pull/6). Further, there is a dependency pull request in https://github.com/tylerjw/panda_moveit_config/pull/4.

rickstaa committed:
Adds missing simple_control_manager dependency (#4)

This commit makes suer that the 'moveit_simple_control_manager' is added
as a dependency. This dependency is needed for running the move_group
script with a real robot.

rickstaa committed:
P/update latest fix melodic noetic diffs (#6)

* Adds rviz_tutorial argument

* Fixes virtual frame world not found warning

rickstaa committed:
Adds 'moveit_visual_tools' as a dependency. (#7)

rickstaa commented:
I created https://github.com/frankaemika/franka_ros/issues/162 to make the Franka Emika team aware of the restrictive hand movement problem. @JafarAbdi please free to add any points I missed or misunderstood in your original explanation to the issue.

rickstaa commented:
> I checked through the commits since the last release and this only backs out one change. There is a PR to add the transport pose to one of the files. This specific file I think we can safely remove, this I explain below.
> 
> This brings us to a point of discussion. There are two sorts of files in this repo that are not generated by moveit_setup_assistant. The first is xacro files that were copied versions of the old robot description. I believe these can be safely removed.
> 
> The next are files that support specific tutorials and were manually added here. An example of that is the `lerp_planning_pipeline.launch.xml` and related changes. I'm not sure what the right approach is for these sorts of files. I think it makes sense to document what these files are so that in the future if someone needs to generate this package to update it they know what to do about those files.
> 
> That or we should move these launch and config files out of this repo and into the tutorials repo to go alongside their respective tutorials. A third option would be to extend the setup assistant to generate these files if they are actually general-purpose launch files that would be useful for other robot setups.
> 
> @rhaschke @v4hn do you have thoughts on how we should best maintain these manually added files? I am working through identifying them and trying to validate the functionality of this PR with tutorials.

Last week I gave this feature another try while keeping the points that @tylerjw  mentioned above in mind. To do this, I first created an orphan branch and created the panda_moveit_config again using the [latest moveit_setup_assistant](https://github.com/ros-planning/moveit/tree/master/moveit_setup_assistant). I did this both to get all the latest changes from the `moveit_setup_assistant` into this repository and to get a sense for which features were missing in the `moveit_setup_assistant`. I then manually compared the new files with those in the [melodic-devel branch](https://github.com/ros-planning/panda_moveit_config/tree/melodic-devel) keeping notes and create pull request where needed. The result of this can be found [here](https://github.com/rickstaa/panda_moveit_config/pull/26).

This pull request is nearly finished, but I might need some help of you guys to clear up some things that were not 100% sure (see https://github.com/rickstaa/panda_moveit_config/pull/26 or the sections below).

After these are finished, I can rebase my orphan branch onto this branch to apply all the required changes (see [this example branch](https://github.com/rickstaa/panda_moveit_config/tree/noetic-devel-update-continued-history)).  Alternatively, I can also rebase what I have currently merged and create new pull requests onto this branch. Let me know what you think is the best solution.

## Pull request that still need to be performed

- [ ] https://github.com/rickstaa/panda_moveit_config/pull/28 I could not add the [trajopt example](https://github.com/ros-planning/moveit_tutorials) yet as the planner has not yet been released. We first have to decide on a strategy on which files and code to add to the current repo.
- [ ] https://github.com/rickstaa/panda_moveit_config/pull/29 I am hesitant to merge this pull request since there appears to be a problem when planning using the STOMP planner (see https://github.com/rickstaa/panda_moveit_config/pull/29#issuecomment-906377648). This problem is explained in this [upstream issue](https://github.com/ros-industrial/industrial_moveit/issues/53). 
- [ ] https://github.com/rickstaa/panda_moveit_config/pull/53 The CHOMP algorithm doesn't seem to be able to control the robot hand. It however works fine with the panda arm. I did not read the internal workings of the CHOMP algorithm, but I think planning for parallel yaw grippers is not supported. I however wanted to verify this before I merge this pull request.
- [ ] I did not merge https://github.com/rickstaa/panda_moveit_config/pull/52 yet since I was not sure if the lerp planner is supposed to work with the panda hand.
- [ ] Add the gazebo simulation. For this I created two different pull request since we first need to decide which approach we want to take (see https://github.com/ros-planning/panda_moveit_config/pull/68#issuecomment-906305362). Contrary to the [melodic-devel](https://github.com/ros-planning/panda_moveit_config/tree/melodic-devel) branch, I now use a `effort_controllers/JointTrajectoryController` controller since the [RobotHWSim](https://github.com/frankaemika/franka_ros/blob/noetic-devel/franka_gazebo/src/franka_hw_sim.cpp) doesn't yet support position hardware interface. I will change to a position controller when https://github.com/frankaemika/franka_ros/pull/156 is merged. The PID's of this controller are not well tuned with the real robot.
    - [ ] https://github.com/rickstaa/panda_moveit_config/pull/39 This commit migrates the codebase such that it uses the gazebo simulation that is found in the [franka_gazebo](https://github.com/frankaemika/franka_ros/tree/noetic-devel/franka_gazebo) package. This needs to be done since the `moveit_setup_assistant` doesn't yet use this simulation. This simulation is more comparible to the real robot. 
    - [ ] https://github.com/rickstaa/panda_moveit_config/pull/40 This commit fixes the gazebo related code that was generated by the `moveit_setup_assistant` such that it works with the latest version of the [franka_ros](https://github.com/frankaemika/franka_ros/) package. This version doesn't use the new  [franka_gazebo](https://github.com/frankaemika/franka_ros/tree/noetic-devel/franka_gazebo) package. Also not merged since we dit not yet decide which gazebo simulation we are going to use.
- [ ] https://github.com/rickstaa/panda_moveit_config/pull/54 I did not yet merge this pull request since I'm not sure if this change is still required with [franka_ros](https://github.com/frankaemika/franka_ros) v0.8.0. I will perform some tests tomorrow.

## Upstream pull requests that are needed.

- [ ] https://github.com/frankaemika/franka_ros/pull/151 needs to be merged to get rid of the ` Link panda_rightfinger has visual geometry but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.` warning.
- [ ] https://github.com/frankaemika/franka_ros/pull/151 needs to be merged to get rid of the `Link panda_leftfinger has visual geometry but no collision geometry` warning.
- [ ] https://github.com/frankaemika/franka_ros/pull/154 needs to be merged or the collision geometries needs to be improved in order to prevent self-collisions.
- [ ] https://github.com/frankaemika/franka_ros/pull/156 and https://github.com/frankaemika/franka_ros/pull/157. Such that we can also use the `joint_position/JointTrajectoryController` controller in the gazebo simulation.
- [ ] https://github.com/frankaemika/franka_ros/pull/145 should be merged for the gazebo simulation to work.

## Things that need to be double-checked

- I currently use the [joint_limit.yaml](https://github.com/rickstaa/panda_moveit_config/blob/melodic-devel/config/joint_limits.yaml) file that is given by the setup assistant. I noticed that the [melodic-devel](https://github.com/rickstaa/panda_moveit_config/blob/melodic-devel/config/joint_limits.yaml) has different values. Maybe somebody knows which values are the right ones? The `moveit_setup_assistant` also sets the `default_velocity_scaling_factor` `default_acceleration_scaling_factor` to `0.1` while this is not done in the [melodic branch](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/joint_limits.yaml).
- The demo_chomp.launch file was added in, but this file is not used in the moveit_tutorials. As a result, this file can be removed or the moveit_tutorial should be edited.
- I currently use the [chomp_planning.yaml](https://github.com/rickstaa/panda_moveit_config/pull/29/files#diff-c8f04138cc92466e9ae5df7d232d7e62832d4776dcbe7c1abbdd7ad62f606bcb) file that is found in the [melodic-devel branch](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/chomp_planning.yaml). This file is however different from [the file](https://github.com/rickstaa/panda_moveit_config/blob/noetic-devel-update/config/chomp_planning.yaml) that is given by the moveit_setup_assistant. Does anybody know which configuration file should be used?
- In the [melodic-devel](https://github.com/ros-planning/panda_moveit_config/tree/melodic-devel) branch the kinematics are passed to the benchmarking node (see https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/launch/run_benchmark_trajopt.launch#L19 and https://github.com/ros-planning/panda_moveit_config/blob/2fd1d2d6ecaccaea72c4799686f7f0143efe0e39/launch/run_benchmark_ompl.launch#L18). In the template of the `moveit_setup_assistant` this was not done. I was not sure if the kinematics needs to be passed to the [RunBenchmark](https://github.com/ros-planning/moveit/blob/master/moveit_ros/benchmarks/src/RunBenchmark.cpp) node or that this is already handled [here](https://github.com/rickstaa/panda_moveit_config/blob/e7e93fc30ec32e4b38fa0af4a8e69279c5a491f8/launch/planning_context.launch#L24-L26).

rhaschke committed:
Merge branch 'melodic-devel' into p/update_latest

rhaschke commented:
Thanks a lot, @rickstaa, for pushing this forward. A new release of panda_moveit_config is long overdue!
I had a look at your branch https://github.com/rickstaa/panda_moveit_config/pull/26 and cleaned it a little bit, following these rules (see [here](https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa) for the result):
- commit messages shouldn't include PR IDs from your fork
- fixups due to changes in MSA behavior were squashed directly into the base commit

Regarding your questions:
- I suggest including examples for other planning pipelines even though these pipelines are currently not operational or released. Otherwise, it will become hard to re-add those files later from scratch.
- For the Gazebo simulation, I suggest to use `franka_gazebo` if that's closer to the real robot. Does that require changes in the tutorials as well?
- If the existing files differ to the MSA-generated ones: Keep the generated ones, if generic files (e.g. planning pipeline settings) are concerned. But keep robot-specific settings like joint limits etc. For those, MSA can only use the defaults from the URDF.
- @Le-Li-17 please have a look at the long list of PRs filed to Franka. We strongly believe that the URDF model released with version 0.8.1 still has severe issues strongly limiting the motion range of the simulated robot. See https://github.com/ros-planning/panda_moveit_config/pull/35#pullrequestreview-390715967

rickstaa commented:
@rhaschke Thanks a lot for helping me implement this new release.

Thanks for rebasing my commits they look a lot cleaner! I will use https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa and adhere to these rules moving forward. Are these rules written down somewhere in a `contributing.md`. Or are they requested when people make a pull request?


> I suggest including examples for other planning pipelines even though these pipelines are currently not operational or released. Otherwise, it will become hard to re-add those files later from scratch.

Are you okay with adding them commented out since otherwise, they throw errors (see https://github.com/rickstaa/panda_moveit_config/blob/f4e1c046d7ca88d9e06f0969ed81cfda62fb2d0b/launch/move_group.launch#L64-L70).

> For the Gazebo simulation, I suggest to use franka_gazebo if that's closer to the real robot. Does that require changes in the tutorials as well?

Good point I did not yet look into this. I check and track all the changes that need to be performed to the `moveit_tutorials` in https://github.com/rickstaa/panda_moveit_config/pull/26.

> @Le-Li-17 please have a look at the long list of PRs filed to Franka. We strongly believe that the URDF model released with version 0.8.1 still has severe issues strongly limiting the motion range of the simulated robot. See Update SRDF for collision model #35 (review)

I'm trying to fix this with @frankaemika. https://github.com/frankaemika/franka_ros/issues/162 is related to this pull request might people want to chip in.

rhaschke commented:
> I will adhere to these rules moving forward. Are these rules written down somewhere in a `contributing.md`?

I just formulated these rules in a common-sense fashion to answer you questions in a generic way. I hope other maintainers agree.

> > I suggest including examples for other planning pipelines even though these pipelines are currently not operational.
> 
> Are you okay with adding them commented out since otherwise, they throw errors.

I wouldn't include them by default into `move_group.launch`, but consider them as optional custom pipelines to be included manually via the `pipeline` argument.

rickstaa commented:
@rhaschke Can I merge the changes in https://github.com/rickstaa/panda_moveit_config/pull/29 (STOMP algorithm) into https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa even though there appears [to be an unsolved upstream issue](https://github.com/rickstaa/panda_moveit_config/pull/29#issuecomment-906377648)?

rickstaa commented:
@rhaschke Do you know if the CHOMP and LERP planners are supposed to work with the panda hand Parallel Jaw gripper?

rhaschke commented:
Big thanks for your work, @rickstaa :tada:
I have pushed the branch to https://github.com/ros-planning/panda_moveit_config/tree/noetic-devel

Could you please open a new issue (e.g. titled `Noetic release`) with all the info in https://github.com/ros-planning/panda_moveit_config/pull/74#issuecomment-947646655?
Closing here.
